File: app/api/knowledge.py
"""知识库上传、公开浏览与检索 API。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, File, Form, Request, UploadFile
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db_session
from app.middleware.visitor_context import get_request_visitor_id
from app.models.knowledge import KnowledgeBaseEntity, KnowledgeDocumentEntity
from app.services.knowledge_service import knowledge_service
from app.services.rag_service import rag_service

router = APIRouter()


class KnowledgeSearchRequest(BaseModel):
    """知识库检索请求体。"""

    query: str = Field(..., min_length=1, description="用户检索问题")
    knowledge_base_id: str | None = Field(default=None, description="单个知识库 ID")
    knowledge_base_ids: list[str] = Field(default_factory=list, description="多个知识库联合检索 ID 列表")
    category: str | None = Field(default=None, description="可选分类过滤")
    top_k: int | None = Field(default=None, ge=1, description="返回结果数量")
    score_threshold: float | None = Field(default=None, ge=0, description="最大距离阈值")
    rewrite_enabled: bool | None = Field(default=None, description="是否启用 query rewrite")


@router.post("/knowledge/upload")
async def upload_knowledge_document(
    request: Request,
    file: UploadFile = File(...),
    name: str = Form(...),
    category: str = Form(...),
    source_type: str = Form(...),
    description: str | None = Form(default=None),
    skill_id: str | None = Form(default=None),
    session: AsyncSession = Depends(get_db_session),
) -> dict[str, Any]:
    """兼容保留的上传接口。"""

    try:
        file_content = await file.read()
        result = await knowledge_service.upload_knowledge_document(
            session,
            name=name,
            category=category,
            source_type=source_type,
            file_name=file.filename or "",
            file_content=file_content,
            description=description,
            skill_id=skill_id,
            visitor_id=get_request_visitor_id(request),
            content_type=file.content_type,
        )
        return _success(result.to_dict())
    finally:
        await file.close()


@router.get("/knowledge/bases")
async def list_public_knowledge_bases(
    category: str | None = None,
    skill_id: str | None = None,
    limit: int = 50,
    session: AsyncSession = Depends(get_db_session),
) -> dict[str, Any]:
    """列出公开可见的启用知识库。"""

    items = await knowledge_service.list_public_knowledge_bases(
        session,
        category=category,
        skill_id=skill_id,
        limit=limit,
    )
    return _success({"items": [_public_knowledge_base_to_dict(item) for item in items]})


@router.get("/knowledge/bases/{knowledge_base_id}")
async def get_public_knowledge_base_detail(
    knowledge_base_id: str,
    session: AsyncSession = Depends(get_db_session),
) -> dict[str, Any]:
    """获取公开可见知识库详情。"""

    knowledge_base, documents = await knowledge_service.get_public_knowledge_base_detail(
        session,
        knowledge_base_id=knowledge_base_id,
    )
    return _success(
        {
            "knowledge_base": _public_knowledge_base_to_dict(knowledge_base),
            "documents": [_public_document_to_dict(document) for document in documents],
        }
    )


@router.post("/knowledge/search")
async def search_knowledge(request: KnowledgeSearchRequest) -> dict[str, Any]:
    """公开知识库检索。"""

    result = await rag_service.search(
        query=request.query,
        knowledge_base_id=request.knowledge_base_id,
        knowledge_base_ids=request.knowledge_base_ids,
        category=request.category,
        top_k=request.top_k,
        score_threshold=request.score_threshold,
        rewrite_enabled=request.rewrite_enabled,
    )
    return _success(result.to_dict())


def _success(data: Any) -> dict[str, Any]:
    return {"code": 200, "message": "success", "data": data}


def _datetime_to_string(value: datetime | None) -> str | None:
    return value.isoformat() if value is not None else None


def _public_knowledge_base_to_dict(entity: KnowledgeBaseEntity) -> dict[str, Any]:
    metadata = dict(entity.metadata_json or {})
    return {
        "id": entity.id,
        "name": entity.name,
        "description": entity.description,
        "category": entity.category,
        "skill_id": entity.skill_id,
        "document_count": int(metadata.get("document_count", 0) or 0),
        "created_at": _datetime_to_string(entity.created_at),
        "updated_at": _datetime_to_string(entity.updated_at),
    }


def _public_document_to_dict(entity: KnowledgeDocumentEntity) -> dict[str, Any]:
    return {
        "id": entity.id,
        "knowledge_base_id": entity.knowledge_base_id,
        "original_file_name": entity.original_file_name,
        "file_extension": entity.file_extension,
        "mime_type": entity.mime_type,
        "file_size": entity.file_size,
        "source_type": entity.source_type,
        "category": entity.category,
        "skill_id": entity.skill_id,
        "index_status": entity.index_status,
        "uploaded_at": _datetime_to_string(entity.uploaded_at),
        "indexed_at": _datetime_to_string(entity.indexed_at),
        "created_at": _datetime_to_string(entity.created_at),
        "updated_at": _datetime_to_string(entity.updated_at),
    }


__all__ = ["router"]

File: frontend/src/types/knowledge.ts
export type KnowledgeUploadResponse = {
  knowledge_base_id: string;
  document_id: string;
  name: string;
  category: string;
  source_type: string;
  skill_id: string | null;
  file_name: string;
  file_size: number;
  index_status: string;
  chunk_count: number;
};

export type AdminKnowledgeBatchUploadItem = {
  file_name: string;
  success: boolean;
  document_id: string;
  index_status: string;
  file_size: number;
  chunk_count: number;
  error_message: string;
};

export type AdminKnowledgeBatchUploadResponse = {
  knowledge_base_id: string;
  total_files: number;
  success_count: number;
  failed_count: number;
  items: AdminKnowledgeBatchUploadItem[];
};

export type KnowledgeSearchRequest = {
  query: string;
  knowledge_base_id?: string | null;
  knowledge_base_ids?: string[];
  category?: string | null;
  top_k?: number | null;
  score_threshold?: number | null;
  rewrite_enabled?: boolean | null;
};

export type RagSearchHit = {
  knowledge_base_id: string;
  document_id: string;
  category: string;
  source_type: string;
  skill_id: string | null;
  file_name: string;
  source: string;
  score: number;
  content: string;
};

export type KnowledgeSearchResponse = {
  query: string;
  rewritten_query: string;
  used_rewrite: boolean;
  top_k: number;
  score_threshold: number | null;
  hits: RagSearchHit[];
  message: string;
};

export type KnowledgeBaseSummaryDto = {
  id: string;
  name: string;
  description: string | null;
  category: string;
  skill_id: string | null;
  document_count: number;
  created_at: string | null;
  updated_at: string | null;
};

export type KnowledgeDocumentSummaryDto = {
  id: string;
  knowledge_base_id: string;
  original_file_name: string;
  file_extension: string;
  mime_type: string;
  file_size: number;
  source_type: string;
  category: string;
  skill_id: string | null;
  index_status: string;
  uploaded_at: string | null;
  indexed_at: string | null;
  created_at: string | null;
  updated_at: string | null;
};

export type KnowledgeBaseListResponse = {
  items: KnowledgeBaseSummaryDto[];
};

export type KnowledgeBaseDetailResponse = {
  knowledge_base: KnowledgeBaseSummaryDto;
  documents: KnowledgeDocumentSummaryDto[];
};

export type AdminKnowledgeBaseDto = {
  id: string;
  name: string;
  description: string | null;
  category: string;
  skill_id: string | null;
  source_type: string;
  status: string;
  is_enabled: boolean;
  document_count: number;
  metadata_json: Record<string, unknown>;
  created_at: string | null;
  updated_at: string | null;
};

export type AdminKnowledgeDocumentDto = {
  id: string;
  knowledge_base_id: string;
  original_file_name: string;
  storage_path: string;
  file_extension: string;
  mime_type: string;
  file_size: number;
  source_type: string;
  category: string;
  skill_id: string | null;
  index_status: string;
  is_enabled: boolean;
  error_message: string | null;
  metadata_json: Record<string, unknown>;
  uploaded_at: string | null;
  indexed_at: string | null;
  created_at: string | null;
  updated_at: string | null;
};

export type AdminKnowledgeBaseListResponse = {
  items: AdminKnowledgeBaseDto[];
};

export type AdminKnowledgeBaseDetailResponse = {
  knowledge_base: AdminKnowledgeBaseDto;
  documents: AdminKnowledgeDocumentDto[];
};

export type AdminKnowledgeBaseCreateRequest = {
  name: string;
  category: string;
  description?: string | null;
  skill_id?: string | null;
};

export type AdminKnowledgeBaseUpdateRequest = Partial<{
  name: string;
  description: string | null;
  category: string;
  skill_id: string | null;
  is_enabled: boolean;
  status: string;
}>;

export type AdminKnowledgeDocumentUpdateRequest = {
  is_enabled: boolean;
};

export type AdminKnowledgeReindexItem = {
  document_id: string;
  file_name: string;
  success: boolean;
  chunk_count: number;
  error_message: string;
};

export type AdminKnowledgeReindexResponse = {
  knowledge_base_id: string;
  total_documents: number;
  success_count: number;
  failed_count: number;
  items: AdminKnowledgeReindexItem[];
};
